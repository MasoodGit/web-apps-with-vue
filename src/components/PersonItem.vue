<template>
  <div>
    <v-list-tile :key="friend.id" avatar ripple @click="like(friend)">
      <v-list-tile-content>
        <v-list-tile-title v-text="fullName"></v-list-tile-title>
      </v-list-tile-content>

      <v-list-tile-action>
        <v-icon :color="friend.fav ? 'red' : 'grey'">favorite</v-icon>
      </v-list-tile-action>
    </v-list-tile>
    <v-divider v-if="last"></v-divider>
  </div>
</template>

<script>
export default {
  props: ["friend", "last"],
  methods: {
    like(friend) {
      friend.fav = !friend.fav;
    }
  },
  computed: {
    fullName() {
      return `${this.friend.firstName} ${this.friend.lastName}`;
    }
  }
};
</script>

<style></style>
