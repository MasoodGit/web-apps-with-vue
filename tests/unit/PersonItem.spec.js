import { mount } from "@vue/test-utils";
import Vuetify from "vuetify";
import PersonItem from "@/components/PersonItem.vue";
import Vue from "vue";

describe("PersonItem", () => {
  beforeEach(() => {
    Vue.use(Vuetify);
  });

  it("should display a Person", () => {
    const wrapper = mount(PersonItem, {
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: true
        },
        last: true
      }
    });
    expect(wrapper.text()).toContain("Raju Gandhi");
  });

  it("should display a divider if its the last friend", () => {
    const wrapper = mount(PersonItem, {
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: true
        },
        last: true
      }
    });
    expect(wrapper.find(".v-divider").exists()).toBe(true);
  });

  it("should NOT display a divider if its NOT the last friend", () => {
    const wrapper = mount(PersonItem, {
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: true
        },
        last: false
      }
    });
    expect(wrapper.find(".v-divider").exists()).toBe(false);
  });

  it("should highlight the favorite icon if the friend is a fav", () => {
    const wrapper = mount(PersonItem, {
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: true
        },
        last: false
      }
    });
    expect(wrapper.find("i.v-icon.red--text").exists()).toBe(true);
  });

  it("should grey out the favorite icon if the friend is NOT a fav", () => {
    const wrapper = mount(PersonItem, {
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: false
        },
        last: false
      }
    });
    expect(wrapper.find("i.v-icon.grey--text").exists()).toBe(true);
    expect(wrapper.find("i").classes()).toContain("grey--text");
  });

  it("should display the full name", () => {
    const wrapper = mount(PersonItem, {
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: false
        },
        last: false
      }
    });
    const vm = wrapper.vm;
    expect(vm.fullName).toBe("Raju Gandhi");
  });

  it("should notifyParent when clicked", () => {
    const wrapper = mount(PersonItem, {
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: false
        },
        last: false
      }
    });

    wrapper.vm.notifyParent = jest.fn();

    wrapper.find("i.v-icon").trigger("click");
    expect(wrapper.vm.notifyParent).toHaveBeenCalledTimes(1);
  });

  it("should emit notifyParent when clicked", () => {
    const wrapper = mount(PersonItem, {
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: false
        },
        last: false
      }
    });

    wrapper.find("i.v-icon").trigger("click");
    const emitted = wrapper.emitted();
    const e = emitted["notify-parent"];
    expect(e).toBeTruthy();
    expect(e.length).toBe(1);
  });

  it("should render correctly", () => {
    const wrapper = mount(PersonItem, {
      propsData: {
        friend: {
          id: 1,
          firstName: "Raju",
          lastName: "Gandhi",
          fav: true
        },
        last: true
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
