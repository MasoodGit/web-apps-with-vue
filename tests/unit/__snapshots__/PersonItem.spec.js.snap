// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonItem should render correctly 1`] = `
<div>
  <div
    role="listitem"
  >
    <a
      class="v-list__tile v-list__tile--link v-list__tile--avatar theme--light"
    >
      <div
        class="v-list__tile__content"
      >
        <div
          class="v-list__tile__title"
        >
          Raju Gandhi
        </div>
      </div>
       
      <div
        class="v-list__tile__action"
      >
        <i
          aria-hidden="true"
          class="v-icon material-icons theme--light red--text"
        >
          favorite
        </i>
      </div>
    </a>
  </div>
   
  <hr
    class="v-divider theme--light"
  />
</div>
`;
